blueprints/APIv3.tex

```diff
@@ -0,0 +1,507 @@
+that violates this restriction is \code{cdns/name/\{\{name\}\}/dnsseckeys/delete}.\\
+This is the standard method to be used by all read-only operations, and as such handlers for this method should generally be accessible to
+users with the "read-only" Role.
+
+\subsubsection{POST}
+POST requests ask the server to process some provided data. Most commonly, in Traffic Ops, this means creating an object based on the serialization
+of said object contained in the request body, but it can also be used virtually whenever no other method is appropriate. When an object \emph{is}
+created, the response body MUST contain a representation of the newly created object.\\
+POST requests do not need to be \emph{idempotent}, unlike PUT requests.
+
+\subsubsection{PUT}
+PUT is used to replace existing data with new data that is provided in the request body.
+\href{https://tools.ietf.org/html/rfc2616#section-9.1.2}{RFC 2616 Section 9.1.2} lists PUT as an "idempotent" request method, which means that
+subsequent identical requests should ensure the same state is maintained on the server. What this means is that a client that PUTs an object
+representation to Traffic Ops expects that if they then GET a representation of that object, do the same PUT again and GET another representation,
+the two retrieved representations should be identical. Effectively, the \code{last\_updated} field that is common to objects in the Traffic Ops API
+violates this, but the other properties of objects - which can actually be defined - generally obey this restriction. In general, fulfilling this
+restriction means that handlers will need to require the entirety of an object be defined in the request body.\\
+This is the method of choice for modifying objects when PATCH cannot be implemented. When an object is modified, the response body MUST contain a
+representation of the object after modification.\\
+While RFC 2616 states that servers MAY create objects for the passed representations if they do not already exist, Traffic Ops API endpoint authors
+should instead use POST handlers for object creation.
+
+\subsubsection{PATCH}
+At the time of this writing, no Traffic Ops API endpoints handle the PATCH request method. PATCH requests that the server's stored data be mutated
+in some way using data provided in the request body. Unlike PUT, PATCH is not \emph{idempotent}, which essentially means that it can be used to
+change only part of a stored object. When an object is modified, the response body MUST contain a representation of the object after modification,
+and that representation SHOULD fully describe the modified object, even the parts that were not modified.\\
+Handlers that implement PATCH in the Traffic Ops API MUST use conditional requests to ensure that race conditions are not a problem, e.g. using
+ETag and If-Match, or If-Not-Modified-Since.\\
```

Comment:
Should be `If-Unmodiifed-Since`. See https://tools.ietf.org/html/rfc7232#section-3.4
Also, consider adding `If-None-Match` and `If-Match`. They're less commonly known, but much better than IMS/IUS. IMO it'd be good to list them as hard requirements as well.

Reply:
I put the hard requirements in a different section, those were just examples. Although, now that it's a hard requirement, maybe I don't need this restriction here?